### PR TITLE
Validate static IPs are not inside allocation pools

### PR DESCRIPTION
**Pull request by tomassedovic**
When entering static IP addresses for the various endpoints, these should not be taken from the allocation pools as that could result in IP conflict during the deployment.

### CODE REVIEW

**Comment by tomassedovic** [comment]
This builds on the #54 pull request. It should be merged after it.

File: ansible-tests/validations/library/validate_network_environment.py [hunk 1]
```diff
@@ -135,6 +139,56 @@ def check_allocation_pools_pairing(filedata, pools):
...
+                ranges_with_conflict = ip_conflicts_with_ranges(ip, pool_ranges)
+                if ranges_with_conflict:
+                    for pool_name, ip_range in ranges_with_conflict:
+                        msg = "IP address {} from {}[{}] is in the {} pool."
+                    errors.append(msg.format(
+                        ip, role, service, pool_name))
+    return errors
+
+
+def ip_conflicts_with_ranges(ip_address, ip_ranges):
```
Comment (mandre) [inline ansible-tests/validations/library/validate_network_environment.py]: It's not clear from the name of the function to guess it filters ranges to the ones conflicting with the given IP address.

I suggest naming it backward, `ranges_conflicting_with_ip` or similar to make it clear it returns ranges.

File: ansible-tests/validations/library/validate_network_environment.py [hunk 2]
```diff
@@ -135,6 +139,56 @@ def check_allocation_pools_pairing(filedata, pools):
...
+    """Check for all conflicts of the IP address conflicts.
+
+    This takes a single IP address and a list of `(pool_name,
+    netenv.IPRange)`s.
+
+    We return all ranges that the IP address conflicts with. This is to
+    improve the final error messages.
+    """
+    results = []
```
Comment (mandre) [inline ansible-tests/validations/library/validate_network_environment.py]: A list comprehension would do wonder here.

File: ansible-tests/validations/library/validate_network_environment.py [hunk 3]
```diff
@@ -135,6 +139,56 @@ def check_allocation_pools_pairing(filedata, pools):
...
+                                allocation_range['end'])
+            pool_ranges.append((pool_name, ip_range))
+    for role, services in six.iteritems(static_ips):
+        for service, ips in six.iteritems(services):
+            for ip in ips:
+                ranges_with_conflict = ip_conflicts_with_ranges(ip, pool_ranges)
+                if ranges_with_conflict:
+                    for pool_name, ip_range in ranges_with_conflict:
+                        msg = "IP address {} from {}[{}] is in the {} pool."
+                    errors.append(msg.format(
```
Comment (mandre) [inline ansible-tests/validations/library/validate_network_environment.py]: Bad indentation

File: testing/static-ip-conflict-with-pool.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
+# Environment file demonstrating how to pre-assign IPs to all node types
```
Comment (mandre) [inline testing/static-ip-conflict-with-pool.yaml:1]: How is this file used exactly? Is it intended for automated testing?

I suggest adding all sorts of errors to the file, like for instance non parsable IP addresses, IP ranges with lower bound higher than the upper bound, etc.

File: ansible-tests/validations/library/validate_network_environment.py (hunk 1 above)
Comment (tomassedovic) [inline ansible-tests/validations/library/validate_network_environment.py]: Good point, will do. And yeah, that's what it does.

File: ansible-tests/validations/library/validate_network_environment.py (hunk 2 above)
Comment (tomassedovic) [inline ansible-tests/validations/library/validate_network_environment.py]: Yeah. I rewrote that bit a few times and forgot to simplify it further.

File: ansible-tests/validations/library/validate_network_environment.py (hunk 3 above)
Comment (tomassedovic) [inline ansible-tests/validations/library/validate_network_environment.py]: Good catch!

File: testing/static-ip-conflict-with-pool.yaml (hunk 4 above)
Comment (tomassedovic) [inline testing/static-ip-conflict-with-pool.yaml:1]: Oh, it's just for manual testing to make sure the validator does actually find the static ip errors like it's supposed to.

There's another file already in there that should be testing the rest of the functionality.

**Pull request merged**